src/hooks/useForm.js

```diff
@@ -0,0 +1,80 @@
+//credit: academind
+
+import { useState, useCallback } from "react";
+
+function useForm(formObj) {
+  const [form, setForm] = useState(formObj);
+
+  function renderFormInputs() {
+    return Object.values(form).map((inputObj) => {
+      const { value, label, errorMessage, valid, renderInput } = inputObj;
+      return renderInput(onInputChange, value, valid, errorMessage, label);
+    });
+  }
+
+  const isInputFieldValid = useCallback(
```

Comment:
**Might be better**

If it’s interesting for you here is a custom `validation hook`.

You should put it in a separate file  `useFormAndValidation.js` in folder `hooks` and then import it to the component that needs validation:

```js

import {useState, useCallback} from 'react';

export function useFormAndValidation() {
  const [ values, setValues ] = useState({});
  const [ errors, setErrors ] = useState({});
  const [ isValid, setIsValid ] = useState(true);

  const handleChange = (e) => {
    const {name, value} = e.target
    setValues({...values, [name]: value });
    setErrors({...errors, [name]: e.target.validationMessage});
    setIsValid(e.target.closest('form').checkValidity());
  };



  const resetForm = useCallback((newValues = {}, newErrors = {}, newIsValid = false) => {
    setValues(newValues);
    setErrors(newErrors);
    setIsValid(newIsValid);
  }, [setValues, setErrors, setIsValid]);

  return { values, handleChange, errors, isValid, resetForm, setValues, setIsValid };
}


```

And only one line of code enables validation:

```js

  const {values, handleChange, errors, isValid, setValues, resetForm} = useFormAndValidation()

```

So, you **don’t need** to make functions for handling inputs, storing values, resetting the form. Everything is already done.